// @graphql-modules/core/dist/commonjs/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var dependency_module_not_found_1 = require("./dependency-module-not-found");
exports.DependencyModuleNotFoundError = dependency_module_not_found_1.DependencyModuleNotFoundError;
var schema_not_valid_1 = require("./schema-not-valid");
exports.SchemaNotValidError = schema_not_valid_1.SchemaNotValidError;
var dependency_module_undefined_1 = require("./dependency-module-undefined");
exports.DependencyModuleUndefinedError = dependency_module_undefined_1.DependencyModuleUndefinedError;
var typedef_not_found_1 = require("./typedef-not-found");
exports.TypeDefNotFoundError = typedef_not_found_1.TypeDefNotFoundError;
var provider_class_not_decorated_1 = require("./provider-class-not-decorated");
exports.ProviderClassNotDecoratedError = provider_class_not_decorated_1.ProviderClassNotDecoratedError;
var module_config_required_1 = require("./module-config-required");
exports.ModuleConfigRequiredError = module_config_required_1.ModuleConfigRequiredError;
